qtgui (and others): Qt6 port

haakov opened the pull request:
Draft PR to keep track of the effort related to https://github.com/gnuradio/gnuradio/issues/7708

haakov committed:
gr-qtgui: Make examples/pyqt_freq_c.py QtPy-compatible

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

maitbot committed:
try qt6 build

Add CMake so that -DTRY_QT6_BUILD=ON will try a Qt6 build.
Otherwise, gnuradio will build using Qt5.
"Try" because without qwt being Qt6 aware, the build will fail.
This is a first step towards Qt6 using gnuradio.

Signed-off-by: A. Maitland Bottoms <bottoms@debian.org>

haakov committed:
qtgui: Delete old/unused QRegExpValidator

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
cmake: Support the Qt6-version of Qwt

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Remove signals/slots from spectrumdisplayform.ui

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Use qtpy in Python examples

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Use qtpy in python/qtgui/

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
runtime/uhd: Don't use "Qt." for widgets

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Use qtpy in uhd_display.py

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
grc-qt: Prefer PyQt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Don't use globalPos() (deprecated)

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

marcusmueller committed:
qtgui: accomodate both Qt5 andf Qt6 when asking for global mouse position

Signed-off-by: Marcus Müller <mmueller@gnuradio.org>

marcusmueller committed:
qtgui: range: use QtWidget in place of Qt

Signed-off-by: Marcus Müller <mmueller@gnuradio.org>

marcusmueller committed:
qtgui: Fix QtCore./Qt. imports

Signed-off-by: Marcus Müller <mmueller@gnuradio.org>

dl1ksv committed:
Fixing some enums that have changed between qt5 qt6

Signed-off-by: Volker Schroer <3470424+dl1ksv@users.noreply.github.com>

dl1ksv committed:
Adding signals and slots to spectrumdisplay, valid for qt

Signed-off-by: Volker Schroer <3470424+dl1ksv@users.noreply.github.com>

dl1ksv committed:
Adopt qtgui-*.yml files for Qt6, modify the genrator template

Signed-off-by: Volker Schroer <3470424+dl1ksv@users.noreply.github.com>

Radar

Signed-off-by: Volker Schroer <3470424+dl1ksv@users.noreply.github.com>

dl1ksv committed:
CPP code generation for qt6

Signed-off-by: Volker Schroer <3470424+dl1ksv@users.noreply.github.com>

marcusmueller commented:
@a17r work in progress. The involved parties have been strapped for time, really.

haakov commented:
With @dl1ksv's last changes, the PR might actually be ready for testing?

It's been a few weeks since I looked at this so I don't have a clear picture of what remains at the moment, but I'll do a rebase now anyways

marcusmueller committed:
qtgui/eyedisplaysform: keep mouse event qt5/6 compatible

Signed-off-by: Marcus Müller <mmueller@gnuradio.org>

marcusmueller committed:
qtgui: Python formatting with black

Signed-off-by: Marcus Müller <mmueller@gnuradio.org>

ZeroChaos- commented:
Gentoo is proceeding with removing qt5.  This will make gnuradio a lot less useful.  Any chance this can get some priority?  Other distros are planning to remove qt5 soon as well.

ZeroChaos- commented:
@marcusmueller I appreciate you and I know you care.  The proposed plan is pretty problematic, as there isn't a release pending for 3.11 (as far as I know) and distros are removing qt5 and that will hurt the usability of gnuradio.  I have zero problem with this plan if you make a 3.11 release soon.  Distros removing qt5 are obviously making big enough changes that they could pull in the 3.11 release, and distros that need to stick to 3.10 also likely aren't removing qt5 as that's a large breaking change.  But without a 3.11 release for folks to move to, newer distros are losing qt support or gnuradio entirely.

Would it maybe be possible to release a 3.11 rc soonish?  Give us qt6 only folks something to latch onto, start testing, reporting bugs, etc?  I don't want to push you for something you aren't ready for, but tagging a beta or rc and asking for testing would likely have a few of us jump on it.  I can always ship a git snapshot instead if you prefer, but that makes support a bit more of a moving target with how active gnuradio development is.  Also having only gentoo users test is likely a small percentage of your user base and a 3.11 tag would get more users :-)

ZeroChaos- commented:
sounds like redhat et al already removed qt5, so sooner is better than later.  Gentoo is trying hard to remove qt5, I expect they may succeed before the end of the month.

If you start dropping a beta/rc as soon as *reasonable* I think it would be best.  Reasonable is defined by you, to avoid getting too many bug reports as the qt6 distros jump on the beta/rc code.  If the end of the month is that point, then do what's best for your user base.  I want it now, but rushing you to release before you are ready won't help either of us, so please give us something as soon as you are comfortable.

marcusmueller commented:
@ZeroChaos- I haven't checked RHEL, but on Fedora you can still install applications that then pull in Qt4 (e.g. the KTimeTracker package does that, currently, although the project itself has moved on to Qt5 by now): I'm personally rather skeptical of anyone touting "we will completely remove Qt5 in the next N months, N < 6", because, uff, that's a few broken things there!

ZeroChaos- commented:
I can't speak for other distros, but Gentoo is aiming to do it this year.

dl1jbe commented:
Ping. As the last comment is nearly half an year old, may I ask about the actual situation here?

hamarituc commented:
> Ping. As the last comment is nearly half an year old, may I ask about the actual situation here?

FWIW: Qt5 deprecation hit me on Gentoo Linux so I wasn't able to use GNU Radio on Gentoo anymore. I rebased this patch set to the current `main` branch and just needed to add some minor tweaks to CMake so GNU Radio finds Qt6 include directories and libraries. You will find the patch set at https://github.com/hamarituc/gentoo-overlay/blob/master/net-wireless/gnuradio/files/gnuradio-qt6.patch. Now GNU Radio compiles against Qt6 under Gentoo and seems to work for me.

dl1jbe commented:
> > Ping. As the last comment is nearly half an year old, may I ask about the actual situation here?
> 
> FWIW: Qt5 deprecation hit me on Gentoo Linux so I wasn't able to use GNU Radio on Gentoo anymore. 

And that was the background of my question. We at Gentoo have deprecated Qt5 for most of the packages and GNU Radio was one of the last - I think in the hope that some progress came out of the discussion above. The Qt5 support for GNU Radio was blocked last week and will be gone on April 30.

I had hoped to trigger some response from the development team with information about the state of coming 3.11 version.

> I rebased this patch set to the current `main` branch and just needed to add some minor tweaks to CMake so GNU Radio finds Qt6 include directories and libraries. You will find the patch set at https://github.com/hamarituc/gentoo-overlay/blob/master/net-wireless/gnuradio/files/gnuradio-qt6.patch. Now GNU Radio compiles against Qt6 under Gentoo and seems to work for me.

That sound interesting. In the meantime I tried it out -  it compiles and seems to be basically functioning.
Lets see what can be done for Gentoo based on that.

marcusmueller commented:
Hi everyone, so state of this: it's work! It's honest, sometimes annoying work. And we have too few people that can acutely spend time on it. One of them is me, and I need to keep the rest of GNU Radio 3 running as well.

We've had a major CMake overhaul which was also long overdue. This invalidates some of the work done here, sadly, but luckily not most of it.

The executive decision on Qt6 support is that we will introduce it as the only Qt supported in GNU Radio 3.11, and that due to API breakage necessary to do so, supporting it on 3.10 is impossible.

I'd be interested in @hamarituc 's work on Qt6 detection, though detection alone isn't the hard part. Making it work with Qwt on different platforms, porting all the python (which can't throw compile-time errors) etc is. 

@dl1jbe note that I have now spent more time dealing with Gentoo pinging me than I had time to work on the thing. Please refrain from pinging about updates. Else I start pinging Gentoo about when it'll start helping us test things on Windows, if they can decide we need to hurry up and follow their Qt5 deprecation. Thank you for understanding!

dl1jbe commented:
Thanks for the feedback and the time invested. Glad to hear that the switch to Qt6 is going forward.

> @dl1jbe note that I have now spent more time dealing with Gentoo pinging me than I had time to work on the thing.

Sorry to hear that. Maybe by looking at the history of the discussion here I missed some other communications from us in the last months which took that much time. In case of that let me apologize.

Anyway, thanks for all your efforts to bring GNU Radio forward to the expected 3.11 version.

marcusmueller commented:
@dl1jbe hope I didn't come across to harshly; I'm still very thankful people (and especially from the Gentoo field) have brought this to my attention. Thing just is that at times, it felt a bit as if I was being badly project-managed (whereas debian's mait just well contributed a patch, as used by @hamarituc) rather than supported :) and that's not how it should be between cooperating FOSS projects.

dl1jbe commented:
> @dl1jbe hope I didn't come across to harshly;  I'm still very thankful people (and especially from the Gentoo field) have brought this to my attention. Thing just is that at times, it felt a bit as if I was being badly project-managed (whereas debian's mait just well contributed a patch, as used by @hamarituc) rather than supported :) and that's not how it should be between cooperating FOSS projects.

Never mind! I know that too much communication (and that includes my 'ping' here),  can be a burden and results sometimes in frustration - been there, done that :-).

Feel free to ask for any help we can do (e.g. for testing in our special environment).

haakov committed:
PyQt5 -> qtpy and a few CMake changes

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
cmake: Remove TRY_QT6_BUILD (Qt5 not supported)

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Add missing includes for Qt6 port

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov commented:
I've made some substantial headway with this PR. It should be ready for testing by the end of the week.

haakov committed:
qtgui: Rename exec_ to exec

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: More exec

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Use <QApplication>, not <qapplication.h>

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
Even more exec_ -> exec

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
github/release: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
grc: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
filter: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
runtime: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
uhd: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
utils: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
zeromq: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
cmake: Update to Qwt Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
readme/docs: Update to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
filter: Fix typos in filter_design

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
Use explicit qtpy imports

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
utils: Resolve Py2->3 and Qt5->6 issues in plot_tools

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
filter: Add explicit imports and exec_->exec in filtdes examples

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Start porting RFNoC fosphor to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
zeromq: Port examples/python/gui.py to Qt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Finish RFNoC fosphor qt6 port (needs testing)

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
zeromq: More qt6 fixes for examples/python/gui.py

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Formatting fixes for qt6 port

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Update binding hash for sink_c

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov commented:
Alright,  with the exception of gr_plot_qt in gr-utils, which uses some Qwt classes that PythonQwt does not expose, this PR is now ready for testing. Will need squashing at some point, obviously.

marcusmueller commented:
I think @haakov is right, `gr_plot_qt` has been broken since the Qt4->Qt5 migration, so nobody's using that and we should (for now?) just remove it from the source tree altogether.

haakov committed:
Replace qtpy with PyQt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
utils: Remove gr_plot_qt (Has been broken for a while)

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
filter: Remove qwt from polezero_plot.py (already uses PyQtGraph)

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
Remove PythonQwt/PyQwt-using scripts

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtpy->PyQt: Correct imports that moved

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
Signal/Slot to pyqtSignal/Slot

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtpy->PyQt: Use scoped enums

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
utils: Remove leftovers from gr_plot_qt

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
Remove more qtpy things, replace with PyQt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
utils: Remove leftovers from pyqt_plot and pyqt_filter

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
grc/utils: Replace deprecated codecs.open() with open()

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
qtgui: Correct vertical range slider direction

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
grcqt: Use FontMetrics horizontalAdvance() instead of width()

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov committed:
grcqt: More scoped enum fixes for PyQt6

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov commented on gr-filter/python/filter/design/filter_design.py:
```diff
@@ -127,15 +127,15 @@ def __init__(self, options, callback=None, restype=""):
         self.gui.action_save.setEnabled(False)
         self.gui.action_open.triggered.connect(self.action_open_dialog)
 
-        self.gui.filterTypeComboBox.currentIndexChanged['const QString&'].connect(
```
Filter design tool needs some testing to make sure I didn't break something here

haakov commented on gr-qtgui/apps/uhd_display.py:
```diff
@@ -19,10 +19,11 @@
+    from PyQt6.QtWidgets import QApplication, QFileDialog, QMainWindow, QWidget
+    from PyQt6 import sip
+except ImportError as e:
+    sys.stderr.write(f"Error: Program requires PyQt/PySide and gr-qtgui: {str(e)}\n")
```
Not PySide

haakov commented on gr-qtgui/examples/pyqt_const_c.py:
```diff
@@ -14,10 +14,10 @@
+    from PyQt6 import QtWidgets, QtGui
+    from PyQt6 import sip
+except ImportError as e:
+    sys.stderr.write(f"Error: Program requires PyQt/PySide and gr-qtgui: {str(e)}\n")
```
Not PySide

haakov commented on gr-qtgui/lib/CMakeLists.txt:
```diff
@@ -8,7 +8,7 @@
 # Setup the QT file generations stuff
 ########################################################################
-
+message(STATUS "QT_DEFAULT_VERSION " ${QT_DEFAULT_MAJOR_VERSION})
```
Remove

haakov commented on gr-qtgui/lib/CMakeLists.txt:
```diff
@@ -69,21 +69,22 @@ set(QTGUI_LIBS
         ${RFNOC_F15_RC_SRCS}
     )
-    gr_find_package(OpenGL)
+    find_package(OpenGL)
```
No reason to depart from `gr_find_package` here, I believe

haakov commented on gr-qtgui/lib/CMakeLists.txt:
```diff
@@ -166,5 +167,5 @@ if(MSVC)
 
 if(BUILD_SHARED_LIBS)
-    gr_library_foo(gnuradio-qtgui qwt::qwt Qt5::Widgets)
+    gr_library_foo(gnuradio-qtgui qwt::qwt Qt6Widgets)
```
I think there's something missing (Qwt?) from this line, requires investigation

haakov commented on grc/gui_qt/components/example_browser.py:
```diff
@@ -16,10 +16,10 @@
 log = logging.getLogger(f"grc.application.{__name__}")
 
 
-class WorkerSignals(QObject):
```
unintended class rename

haakov commented on grc/gui_qt/components/example_browser.py:
```diff
@@ -33,11 +33,11 @@ def __init__(self, fn, *args, **kwargs):
         self.args = args
         self.kwargs = kwargs
-        self.signals = WorkerSignals()
+        self.signals = WorkerpyqtSignals()
```
Also unintended

haakov commented on grc/main.py:
```diff
@@ -185,12 +185,16 @@ def run_qt(args, log):
 def main():
+    # Prefer PyQt6, but accept other flavors
+    if os.environ.get("QT_API") is None:
+        os.environ["QT_API"] = "pyqt6"
```
remove

haakov commented on README.md:
````diff
@@ -45,9 +45,9 @@ energy particle physics, astrophysics, radio astronomy and more!
 ```bash
-    # Ensure PyQt5 and other dependencies are installed
+    # Ensure PyQt6 and other dependencies are installed
     pip install QDarkStyle qtpy
````
remove

a17r commented on gr-qtgui/lib/CMakeLists.txt:
```diff
@@ -166,5 +167,5 @@ if(MSVC)
 
 if(BUILD_SHARED_LIBS)
-    gr_library_foo(gnuradio-qtgui qwt::qwt Qt5::Widgets)
+    gr_library_foo(gnuradio-qtgui qwt::qwt Qt6Widgets)
```
```suggestion
    gr_library_foo(gnuradio-qtgui qwt::qwt Qt6::Widgets)
```

haakov replied:
Fixed, thanks

haakov committed:
qt5 -> qt6 minor fixes

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

haakov commented:
Note on what is done in this PR (and why it's so big):
- PyQt5 allows unscoped enums, while PyQt6 does not. This means that Qt.Horizontal must be changed to Qt.Orientation.Horizontal, for example.
- exec_() is deprecated in Qt6 -> exec()
- I have removed qtpy from GRC Qt now that we no longer need to support PyQt5
- I have switched to explicit imports for PyQt, except in the Mako templates 
- I don’t believe I have the hardware/license to test the RFNoC F15 code, so that part is untested at the moment
- I have removed everything that uses PyQwt/PythonQwt, but not C++ Qwt

haakov commented on .github/workflows/make-test.yml:
```diff
@@ -60,7 +60,7 @@ jobs:
         # -----------------------------------
   linux-docker:
   # All of these shall depend on the formatting check (needs: check-formatting)
-    needs: [check-hashes, check-formatting, check-python-formatting]
```
?

haakov committed:
ci: Update Fedora images (43 3.10, 44 3.11)

Signed-off-by: Håkon Vågsether <hakon.vagsether@gmail.com>

dl1jbe committed:
Adapt for PyQt and Qt Version 6.11

fedepell commented:
First of all I wanted to chime in and thank you a lot for this work!

I have started testing this on AlmaLinux 10.2 (so RPM based, Qt 6.10.1 there, created a RPM in our builder) and built using main+this patch with no problems!

I have then started using many examples (with base adapted for Qt6 as expected) and saw no problem whatsoever until now, all working good!

I will try to push the team using GNUradio at my org to test it now in detail with real use cases and let you know!

Many thanks again, really looking forward to this merged!

smunaut commented:
FWIW I just built this and it works fine.
I was also able to port fosphor OOT to use Qt6 and that integrates fine as well.

marcusmueller commented:
hey, @dl1jbe, as we near merging of this PR: I see a commit of yours (thank you!!) has no `Signed-Off-By: yourname <youremail>`. We generally need that to avoid having code that we shouldn't have in-tree; I can, however, rewrite the commit to say that *I* know it's really your code that you're allowed to contribute, if you could explicitly confirm that here :)

marcusmueller commented:
Conda doesn't find Qt6, but this is not a blocker for this PR; we need to repair it, either way.

marcusmueller commented:
Mingw-w64 fails due to qa_fastnoise, as others. This is something that's not due to this PR (and I have a fix in the Pipeline)

dl1jbe commented:
> hey, @dl1jbe, as we near merging of this PR: I see a commit of yours (thank you!!) has no `Signed-Off-By: yourname <youremail>`.

Sorry, missed that.

> I can, however, rewrite the commit to say that _I_ know it's really your code that you're allowed to contribute, if you could explicitly confirm that here :)

Yes, that seems to be the easiest way. Thanks.

And yes I confirm that the commit 'Adapt for PyQt and Qt Version 6.11' was provided by myself.

marcusmueller commented:
@dl1be thanks!

marcusmueller commented:
Doesn't look like I can fix the conda recipe.

marcusmueller reviewed (approved):
Reviewed (all but the .grc files) line-by-line

All the removals were a clear "oops, didn't realize we were schlepping that around", thanks for the drive-by bug fixes (integer vs float div, QPoints etc), and, most importantly:

thanks to everyone involved getting this done!

kgarrels commented:
Hi,

I tried to build this with conda on macos tahoe, it seems there are some homebrew references in the way:
```
-- ######################################################
-- # Gnuradio disabled components                        
-- ######################################################
--   * doxygen-documentation
--   * JSON/YAML config blocks
-- 
-- Using install prefix: /Users/devel/micromamba/envs/grdev
-- Building for version: v3.11.0.0git-1168-g61f074b6 / 3.11.0git
-- Configuring done (14.2s)
CMake Error in grc/CMakeLists.txt:
  Imported target "Qt6::PrintSupport" includes non-existent path

    "/opt/homebrew/include/qt6/QtPrintSupport"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in grc/CMakeLists.txt:
  Imported target "Qt6::PrintSupport" includes non-existent path

    "/opt/homebrew/include/qt6/QtPrintSupport"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in gr-qtgui/lib/CMakeLists.txt:
  Imported target "Qt6::Widgets" includes non-existent path

    "/opt/homebrew/include/qt6/QtWidgets"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



CMake Error in gr-qtgui/lib/CMakeLists.txt:
  Imported target "Qt6::Widgets" includes non-existent path

    "/opt/homebrew/include/qt6/QtWidgets"

  in its INTERFACE_INCLUDE_DIRECTORIES.  Possible reasons include:

  * The path was deleted, renamed, or moved to another location.

  * An install or uninstall procedure did not complete successfully.

  * The installation package was faulty and references files it does not
  provide.



-- Generating done (0.5s)
CMake Generate step failed.  Build files cannot be regenerated correctly.
```
